### ISSUE TITLE
Making Unlimiformer work with decoder models (specifically LLaMA)

### ISSUE THREAD

**Issue by StrangeTcy** [abertsch72/unlimiformer#15]
[This](https://github.com/abertsch72/unlimiformer/issues/2) is a related issue, and in this issue I'd like to get technical.
I've been working on trying to adapt Unlimiformer to work with LLaMA for a while now, and it has come down to two main issues (so far): naming and architecture.
As an example of naming, in `Unlimiformer.create_key_value` the original attention is calculated from encoder_attn: https://github.com/abertsch72/unlimiformer/blob/5b534d1532246da8ef3fb02bdfff41aa853b12de/src/unlimiformer.py#L751
I had to replace this with `self_attn` in `UnlimiformerLlama`, since it's the only kind of attention LLaMA has.

And the issue with the architecture came up when trying to calculate the key:
https://github.com/abertsch72/unlimiformer/blob/5b534d1532246da8ef3fb02bdfff41aa853b12de/src/unlimiformer.py#LL753C30-L753C30
Neither LLaMA nor GPT has any encoders, and so it has no `encoder_hidden_states`, and `attention.k_proj` doesn't accept a `None` value it defaults to.

There's probably some simple solution to this, some other state that should be used in decoder-only models in this case, but I haven't figured it out yet, neither on my own nor re-reading the original paper.

So, any pointers would be welcome.

**Comment by SharkWipf** [comment]
Can't add any useful insights other than that I'll be watching this issue closely. 
I still feel like Unlimiformer can be the biggest leap in decoder models we've seen to date, and it pains me to see it getting so little attention, snowed in by hundreds of other whitepapers and people missing its potential.  
From what I understand (which, admittedly, is not much), Unlimiformer effectively beats approaches like Facebook's MEGABYTE in quality, while not being mutually exclusive either. 
Aside from "just" long context windows, it should also allow per-character tokenization, which, while it'd cut into the total text size supported significantly, might allow for much better models trained from scratch, and still have longer context size than current models on top.
While I can't add any useful insights on how to add decoder-only support, I hope I can get more people motivated in helping out by highlighting its potential.

**Comment by urialon** [comment]
Hi @StrangeTcy and @SharkWipf ,
Thank you for your interest in Unlimiformer and for your kind words! (@SharkWipf please don't delete your post!)

The huge attention from the community convinced us that we should implement support for decoder-only models :-D 

I haven't started working on it yet, so if any of you guys manage to make progress we would love to get contributions, but I think that the major difference is that decoder-only models would **require an index per layer**, whereas in encoder-decoder models we could have a single index for the `encoder_hidden_states`. The difference is that in encoder-decoder models - **all decoder layers attend to the last encoder layer's output** (so we could have a single index), while in decoder-only models - every layer attends only to itself.

In other words, the implementation would require keeping `num_layers` indexes. However, I do think that we will be able to do the attention trick that we describe [in Equation 2 in the paper](https://arxiv.org/pdf/2305.01625.pdf):
<img width="245" alt="image" src="https://github.com/abertsch72/unlimiformer/assets/15002544/de24a4f0-a4d5-48a1-8db9-baee23de8982">

which will allow sharing the layer-specific index across all heads. So we could have `num_layers` indexes, rather than `num_layers * num_heads`.

What do you think? does it make sense?

**Comment by StrangeTcy** [comment]
Thanks, I'll have to re-read the paper and think about the whole attention reformulation for a while...

Let this be a placeholder for my actual thoughts...:
<THOUGHTS HERE>

**Comment by urialon** [comment]
No worries @SharkWipf 😄

**Comment by urialon** [comment]
Hey @SharkWipf and @StrangeTcy ,

Unlimiformer now supports Llama and Llama-2 (and all their derivatives)!

Check it out, and let us know if you have any questions!
https://github.com/abertsch72/unlimiformer#august-2023---unlimiformer-now-supports-llama-2-and-all-its-derivatives

Best,
Uri

**Comment by AshwinRamachandran2002** [comment]
Hi, could you provide some details as to how you added support for decoder only models such as Llama?
